compiler/semantic/analyzer.go

```diff
@@ -14,28 +14,104 @@ import (
 // to DAG form, resolving syntax ambiguities, and performing constant propagation.
 // After semantic analysis, the DAG is ready for either optimization or compilation.
 func Analyze(ctx context.Context, seq ast.Seq, source *data.Source, head *lakeparse.Commitish) (dag.Seq, error) {
-	c := newAnalyzer(ctx, source, head)
-	return c.check(seq)
+	a := newAnalyzer(ctx, source, head)
+	s, err := a.semSeq(seq)
+	if err != nil {
+		return nil, err
+	}
+	if err := a.checkOpCycle(); err != nil {
+		return nil, err
+	}
+	return s, nil
 }
 
 type analyzer struct {
-	ctx    context.Context
-	source *data.Source
-	scope  *Scope
-	head   *lakeparse.Commitish
-	zctx   *zed.Context
+	ctx       context.Context
+	head      *lakeparse.Commitish
+	opDeclMap map[*dag.UserOp]*opDecl
+	source    *data.Source
+	scope     *Scope
+	zctx      *zed.Context
+
+	// opDecl is the current operator declaration being analyzed.
+	opDecl *opDecl
 }
 
 func newAnalyzer(ctx context.Context, source *data.Source, head *lakeparse.Commitish) *analyzer {
 	return &analyzer{
-		ctx:    ctx,
-		source: source,
-		scope:  NewScope(),
-		head:   head,
-		zctx:   zed.NewContext(),
+		ctx:       ctx,
+		head:      head,
+		opDeclMap: make(map[*dag.UserOp]*opDecl),
+		source:    source,
+		scope:     NewScope(nil),
+		zctx:      zed.NewContext(),
 	}
 }
 
-func (c *analyzer) check(seq ast.Seq) (dag.Seq, error) {
-	return c.semSeq(seq)
+func (a *analyzer) enterScope() {
+	a.scope = NewScope(a.scope)
+}
+
+func (a *analyzer) exitScope() {
+	a.scope = a.scope.parent
+}
+
+type opDecl struct {
+	op   *dag.UserOp
+	deps []*dag.UserOp
+}
+
+func (a *analyzer) checkOpCycle() error {
+	visited := make(map[*dag.UserOp]bool)
+	onStack := make(map[*dag.UserOp]bool)
+	if p := a.checkScopeOpCycle(a.scope, visited, onStack); p != nil {
+		return opCycleError(p)
+	}
+	return nil
+}
+
+func (a *analyzer) checkScopeOpCycle(scope *Scope, visited, onStack map[*dag.UserOp]bool) []*dag.UserOp {
+	for _, e := range scope.sortedEntries() {
+		if op, ok := e.ref.(*dag.UserOp); ok {
+			if p := a.isCyclic(op, visited, onStack, nil); p != nil {
+				return p
+			}
+		}
+	}
+	for _, child := range scope.children {
+		if p := a.checkScopeOpCycle(child, visited, onStack); p != nil {
+			return p
+		}
+	}
+	return nil
+}
+
+func (a *analyzer) isCyclic(op *dag.UserOp, visited, onStack map[*dag.UserOp]bool, path []*dag.UserOp) []*dag.UserOp {
+	path = append(path, op)
+	visited[op], onStack[op] = true, true
+	for _, neighbor := range a.opDeclMap[op].deps {
+		if !visited[neighbor] {
+			if p := a.isCyclic(neighbor, visited, onStack, path); p != nil {
+				return p
+			}
+		} else if onStack[neighbor] {
+			path = append(path, neighbor)
+			return path
```

Comment:
You don't need `onStack` since `path` holds the current stack. And given that, `stack` would be a better name for it than `path`.
```suggestion
		} else if slices.Contains(path, neighbor) {
			return append(path, neighbor)
```

Reply:
Great point.